lib/lookup/src/lookup_v2/jit.rs

```diff
@@ -0,0 +1,216 @@
+                        JitState::Continue => None,
+                        JitState::Dot => None,
+                        JitState::IndexStart => Some(BorrowedSegment::Invalid),
+                        JitState::Index { .. } => Some(BorrowedSegment::Invalid),
+                        JitState::Field { start } => {
+                            Some(BorrowedSegment::Field(&self.path[start..]))
+                        }
+                        JitState::Quote { .. } => Some(BorrowedSegment::Invalid),
+                        JitState::End => None,
+                    };
+                    self.state = JitState::End;
+                    return result;
+                }
+                Some((index, c)) => {
+                    let (result, state) = match self.state {
+                        JitState::Start | JitState::Continue => match c {
+                            '.' => (None, JitState::Dot),
+                            'A'..='Z' | 'a'..='z' | '_' | '0'..='9' => {
+                                (None, JitState::Field { start: index })
+                            }
+                            '[' => (None, JitState::IndexStart),
+                            '\"' => (None, JitState::Quote { start: index + 1 }),
+                            _ => (Some(Some(BorrowedSegment::Invalid)), JitState::End),
+                        },
+                        JitState::Dot => match c {
+                            'A'..='Z' | 'a'..='z' | '_' | '0'..='9' => {
+                                (None, JitState::Field { start: index })
+                            }
+                            '\"' => (None, JitState::Quote { start: index + 1 }),
+                            _ => (Some(Some(BorrowedSegment::Invalid)), JitState::End),
+                        },
+                        JitState::Field { start } => match c {
+                            'A'..='Z' | 'a'..='z' | '_' | '0'..='9' => {
+                                (None, JitState::Field { start })
+                            }
+                            '.' => (
+                                Some(Some(BorrowedSegment::Field(&self.path[start..index]))),
+                                JitState::Dot,
+                            ),
+                            '[' => (
+                                Some(Some(BorrowedSegment::Field(&self.path[start..index]))),
+                                JitState::IndexStart,
+                            ),
+                            _ => (Some(Some(BorrowedSegment::Invalid)), JitState::End),
+                        },
+                        JitState::Quote { start } => match c {
+                            '\"' => (
+                                Some(Some(BorrowedSegment::Field(&self.path[start..index]))),
+                                JitState::Continue,
+                            ),
+                            _ => (None, JitState::Quote { start }),
+                        },
```

Comment:
It doesn't appear that this is supporting escaped quotes inside a quoted path, e.g. `.foo."bar\"baz"`.

Reply:
I seemed to have missed that. That's very unfortunate since single character escaping requires a memory allocation to support. I'll think about this for a bit...

Reply:
That's correct. It is unfortunate. It's one area that caused the need for allocation before (along with supporting coalesced paths).

Reply:
I can at least implement it in a way where the allocation doesn't happen unless it's actually needed, which I think should be enough since it's probably rare to need this.

Reply:
I ended up having to switch `BorrowedSegment::Field` to use a `Cow`, but it doesn't do any allocations unless a segment actually contains character escaping. I didn't notice any perf regressions for paths that don't use this (which is likely the majority of them)